### PR TITLE
fix(gtp): correct expert gradient scaling for GTP/EGTP mismatch (non-per-token-loss)

### LINKED ISSUE #7164: Support Qwen3.5-VL with GTP

### PR DESCRIPTION
**Pull request by fanshiqing**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

- Bug: expert_gradient_scaling_factor reused dp_cp_group.size() (shrunk by GTP) as
      the prescale for expert params, but experts only ever recover the EGTP share of
      that shrinkage back, not the full GTP share -- leaving a gtp/egtp scaling error
      whenever the two differ.

- Fix: fold an egtp/gtp correction into the prescale, applied unconditionally
      (needed for both gtp>egtp and gtp<egtp).
- Tests: synthetic + real-MoE regression coverage for both mismatch directions.

### EGTP=1: the fix makes `grad-norm` and `params-norm` healthy again

image

✅ Is PR-7165 working? Yes, unambiguously. Both metrics that were outside their own A/A band are now inside.
✅ What got healthy — grad norm 8.4x outside → 1.005, and param norm 8.7x outside → −0.003 %. These are the two the fix targets.
✅ What else improved — lm +0.0202 → +0.0007 % and mtp_2 +0.0393 → +0.0016 %, each tightening by more than an order of magnitude.

### EGTP=4: the fix shrink the `grad-norm` and `params-norm` to healthy state too, no spike anymore

image

## Issue tracking

This MR is based on @xuwchen 's fix: https://github.com/NVIDIA/Megatron-LM/pull/7160

Linked issue: Related to #7164 

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### CODE REVIEW

**Review by kvareddy — APPROVED** [review]

File: tests/unit_tests/generalized_tensor_parallel/test_gtp_grad_correctness.py [hunk 1]
```diff
@@ -377,11 +382,12 @@ def _worker_distopt(rank, world_size, port):
...
             f"\n[distopt grad-norm] baseline={base_gn:.6f}  GTP_remat={gtp_gn:.6f}  "
-            f"ratio={ratio:.4f}",
+            f"ratio={ratio:.6f}",
             flush=True,
         )
-        # Same model, same data, gradients proven equal -> grad-norm must match.
-        torch.testing.assert_close(torch.tensor(gtp_gn), torch.tensor(base_gn), atol=0, rtol=3e-2)
+        # Same model, same data, gradients proven equal -> grad-norm must match (bf16 Adam
+        # noise floor is ~4e-5 here).
+        torch.testing.assert_close(torch.tensor(gtp_gn), torch.tensor(base_gn), atol=0, rtol=1e-4)
```
[the remark this answers is not in the document]
Comment (fanshiqing) [reply tests/unit_tests/generalized_tensor_parallel/test_gtp_grad_correctness.py:390]: it's fine for the current changes

File: tests/unit_tests/generalized_tensor_parallel/test_gtp_grad_correctness.py [hunk 2]
```diff
@@ -516,10 +541,114 @@ def _worker_moe_distopt(rank, world_size, port):
         ratio = moe_gn / max(base_gn, 1e-12)
         print(
             f"\n[moe distopt grad-norm] baseline={base_gn:.6f}  GTP_remat={moe_gn:.6f}  "
-            f"ratio={ratio:.4f}",
+            f"ratio={ratio:.6f}",
             flush=True,
         )
-        torch.testing.assert_close(torch.tensor(moe_gn), torch.tensor(base_gn), atol=0, rtol=3e-2)
+        # bf16 Adam noise floor is ~6e-5 here.
+        torch.testing.assert_close(torch.tensor(moe_gn), torch.tensor(base_gn), atol=0, rtol=1e-4)
```
[the remark this answers is not in the document]
Comment (fanshiqing) [reply tests/unit_tests/generalized_tensor_parallel/test_gtp_grad_correctness.py:548]: it's fine for the current changes, already verified locally

**Review by Phlip79 — APPROVED** [review]

File: megatron/core/distributed/distributed_data_parallel.py [hunk 3]
```diff
@@ -268,14 +268,28 @@ def __init__(
             #   1. Scale gradients by 1/dp_size before reduction
             #   2. Do sum reduction across data parallel ranks
             #   3. Final result is scaled by 1/dp_size as desired
+            #
```
Comment (yashaswikarnati) [inline megatron/core/distributed/distributed_data_parallel.py:275]: just a minor nit; may be this comment can be concisely written

**Review by yashaswikarnati — APPROVED** [review]

File: megatron/core/distributed/distributed_data_parallel.py (hunk 3 above)
Comment (fanshiqing) [reply megatron/core/distributed/distributed_data_parallel.py:275]: rewrite again to make it easier to understand
